File: src/main/java/audioWfc/wfc/constraints/ChordPrototypeOnlyFollowedByConstraint.java
package audioWfc.wfc.constraints;

import audioWfc.musicTheory.chords.Chord;
import audioWfc.wfc.HigherValues;
import audioWfc.wfc.Tile;
import audioWfc.wfc.constraints.concepts.HardConstraint;
import audioWfc.wfc.grabbers.Grabber;
import audioWfc.wfc.hierarchy.prototypes.ChordPrototype;
import audioWfc.wfc.hierarchy.prototypes.Chordesque;
import audioWfc.wfc.hierarchy.prototypes.Section;

import java.util.Set;

public class ChordPrototypeOnlyFollowedByConstraint implements HardConstraint<Chordesque> {
    private String chordPrototypeName;
    private Grabber<Set<Chord>> followingChordsGrabber;

    public ChordPrototypeOnlyFollowedByConstraint(String chordPrototypeName, Grabber<Set<Chord>> followingChordsGrabber) {
        this.chordPrototypeName = chordPrototypeName;
        this.followingChordsGrabber = followingChordsGrabber;
    }

    @Override
    public String name() {
        return "Section is followed by";
    }

    @Override
    public String configText() {
        return followingChordsGrabber.configText();
    }

    @Override
    public boolean check(Tile<Chordesque> tile, HigherValues higherValues) {
        Chordesque chordesque = tile.getValue();
        boolean out = true;

        Tile<Chordesque> prevTile = tile.getPrev();
        if(prevTile.isCollapsed()){
            out &= check(prevTile.getValue(), chordesque, higherValues);
        }

        Tile<Chordesque> nextTile = tile.getNext();
        if(prevTile.isCollapsed()){
            out &= check(chordesque, nextTile.getValue(), higherValues);
        }

        return out;
    }

    private boolean check(Chordesque a, Chordesque b, HigherValues higherValues){
        if(!(a instanceof ChordPrototype c))return true;
        if (!c.getName().equals(chordPrototypeName)) return true;
        Set<Chord> followingChords = followingChordsGrabber.grab(higherValues);
        return followingChords.contains(b.getValue());
    }
}

File: vite-project/src/wfc/constraints/ChordPrototypeOnlyFollowedByConstraint.ts
import { Chord } from "../../music_theory/Chord"
import { HigherValues } from "../HigherValues"
import { Tile } from "../Tile"
import { Chordesque, ChordPrototype } from "../hierarchy/prototypes"

export class ChordPrototypeOnlyFollowedByConstraint implements HardConstraint<Chordesque> {
	private chordPrototypeName: string
	private grabber: Grabber<Set<Chord>>

	name = "Chord Prototype Only Followed By"
	configText = () => `Chord Prototype: ${this.chordPrototypeName}, Chord Set: ${this.grabber.configText()}`

	constructor(chordPrototypeName: string, grabber: Grabber<Set<Chord>>) {
		this.chordPrototypeName = chordPrototypeName
		this.grabber = grabber
	}

	check(tile: Tile<Chordesque>, higherValues: HigherValues): boolean {
		const chordesque = tile.getValue()
		const prev = tile.getPrev()
		const next = tile.getNext()

		let out = true
		if(prev.isCollapsed()) out = out && this.checkPair(prev.getValue(), chordesque, higherValues)
		if(next.isCollapsed()) out = out && this.checkPair(chordesque, next.getValue(), higherValues)
		return out
	}

	private checkPair(first: Chordesque, second: Chordesque, higherValues: HigherValues): boolean {
		if(!(first instanceof ChordPrototype)) return true
		if(first.getName() !== this.chordPrototypeName) return true
		const chordSet = this.grabber.grab(higherValues)
		return chordSet.has(second.getChord())
	}
}